### PR TITLE
Waveform fitting cleanup

### PR DESCRIPTION
**Pull request by pinkenburg**
[comment]: <> (Please tell us something about this pull request)

## Types of changes
[comment]: <> ( What types of changes does your code introduce? Put an `x` in all the boxes that apply: )
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work for users)
- [ ] Requiring change in macros repository (Please provide links to the macros pull request in the last section)
- [x] I am a member of [GitHub organization of sPHENIX Collaboration](https://github.com/orgs/sPHENIX-Collaboration/people), EIC, or ECCE (contact Chris Pinkenburg to join)

## What kind of change does this PR introduce? (Bug fix, feature, ...)

[comment]: <> ( What does this PR do? Linking to talk in software meeting encouraged )
This PR is a cleanup of the g4waveformsim package, no functionality changes

## TODOs (if applicable)

[comment]: <> ( In case this is a draft PR, e.g. for running checks using Jenkins, please make the pull request as a draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/  )

## Motivation / Context
This PR is a maintenance/refactoring cleanup of **g4waveformsim** driven by Valgrind-detected memory leak issues and “code quality/memory safety” improvements. The stated goal is **no intended physics/reconstruction behavior change**, but several internal refactors do touch waveform/calibration setup paths.

## Key changes
- **Memory/resource cleanup**
  - Updated `CaloWaveformSim` destructor to additionally delete calibration-related `CDBTTree` pointers and the waveform template (`h_template`), not just the RNG.
- **`InitRun` refactor (calibration + setup)**
  - Waveform template loading refactored to `TFile::Open()` + `GetObject("hpwaveform", h_template)`.
  - `m_runNumber` now comes from `recoConsts::instance()->get_IntFlag("RUNNUMBER")` instead of reading the `EventHeader`.
  - Detector initialization now includes an explicit **`HCALOUT`** branch and improved invalid-detector error handling.
  - Calibration resolution rewritten:
    - Direct URLs resolved via `CDBInterface::instance()->getUrl(...)`.
    - Conditionally constructs `CDBTTree` objects for **data**, optional **MC energy**, and optional **time** calibrations.
    - Adds verbosity-controlled logging for selected calibration selections.
    - Exits early when required calibrations are unavailable (notably when data/time calibration is enabled but missing).
- **Event processing refactors**
  - Introduced/used a per-tower memoization cache (`tbt_smear`) for Gaussian smearing factors keyed by an encoded identifier.
  - Waveform saturation after noise/pedestal injection now uses `std::clamp` to enforce samples in **[0, 16383]**.
- **API/build system cleanup**
  - Removed `CaloWaveformSim::End(PHCompositeNode*)`.
  - Header/build dependency cleanup (include reorganization; Makefile uses `-isystem` for an offline include path; link list reordered/trimmed).

## Potential risk areas
- **Reconstruction/calibration behavior**: The calibration setup logic was substantially rewritten (direct-URL handling, early-exit conditions, selection of data/MC/time trees). Even if intended to be equivalent, it should be validated for common run configurations.
- **Waveform output differences**: Switching saturation logic to `std::clamp` and changing template loading method can lead to subtle differences at boundaries/edge cases.
- **Data products / IO shape**: Adding an explicit `HCALOUT` branch may affect downstream consumers expecting specific branch sets.
- **Thread-safety / performance**: The new `tbt_smear` memoization cache could have implications if instances are used in multithreaded contexts (ensure it’s per-instance and not shared unsafely). Performance should generally improve (less repeated sampling), but verify under realistic workloads.
- **Memory correctness**: Destructor changes are intended to eliminate leaks, but ownership/double-free risks should be checked if any of those pointers are also owned elsewhere.

## Possible future improvements
- Factor calibration selection (data vs MC energy vs time, directURL vs CDB name) into a small helper/config struct to reduce conditional complexity.
- Add targeted integration tests covering:
  - direct-URL calibration vs CDB-name calibration,
  - “missing calibration” early-exit behavior,
  - waveform saturation boundary conditions.
- Add explicit notes/assertions for multithreading assumptions and cache lifetime.

**Note:** AI-generated summaries can miss nuances—please use best judgment and verify behavior against your expected physics/reconstruction workflows before merging.

### CODE REVIEW

File: simulation/g4simulation/g4waveformsim/CaloWaveformSim.cc [hunk 1]
```diff
@@ -128,92 +130,146 @@ int CaloWaveformSim::InitRun(PHCompositeNode *topNode)
...
   {
-    cdbttree = new CDBTTree(url);
+    cdbttree = new CDBTTree(m_directURL);
   }
   else
   {
-    std::cerr << "CaloWaveformSim::InitRun No data calibration for " << m_calibName << std::endl;
+    std::cout << Name() << ": CaloWaveformSim::InitRun No data calibration for " << m_calibName << std::endl;
     exit(1);
   }
```
[the remark this answers is not in the document]
Comment (pinkenburg) [reply simulation/g4simulation/g4waveformsim/CaloWaveformSim.cc:180]: ran valgrind which confirmed the leaks, implemented the suggestion (also deleted the TProfile)
